add int4 packed gemm support on CPU device

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118056
* __->__ #117475

This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
add int4 packed gemm support on CPU device

add int4 packed gemm on avx2

refine blocking on K

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,541 @@
+    // output:
+    //   scale: {s0, s1, s2, ..., s15}
+    //   zero:  {z0, z1, z2, ..., z15}
+    scale[i] = _mm512_mask_permutex2var_ps(a, 0xffff, idx1, b);
+    zero[i] = _mm512_mask_permutex2var_ps(a, 0xffff, idx2, b);
```
right now, gptfast did not provide an interface to change the format of `scales_and_zeros`, Also I have tried to pre-pack `scales_and_zeros` so as to skip the transpose here, turns out it did not much any difference ...

mingfeima commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,541 @@
+  // 32, 64, 128, 256
+  const int BLOCK_K = qGroupSize;
+
+  const int MB = (M + BLOCK_M - 1) / BLOCK_M;
```
Sure, will add this path later on.
The default benchmark from gpt fast has small prompt length so I planned to do it  later because this won't make much difference from the current benchmark.

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `28.36 sec total, 7.05 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `11.91 sec total, 16.80 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@leslie-fang-intel @jgong5 any idea why CI failed with `RuntimeError: Dynamo is not supported on Python 3.12+` ?

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented:
> @leslie-fang-intel @jgong5 any idea why CI failed with `RuntimeError: Dynamo is not supported on Python 3.12+` ?

Seems like an environment issue. @malfet any idea how to avoid it (https://github.com/pytorch/pytorch/actions/runs/8092102563/job/22113262294?pr=117475)?

mingfeima commented:
> > @leslie-fang-intel @jgong5 any idea why CI failed with `RuntimeError: Dynamo is not supported on Python 3.12+` ?
> 
> Seems like an environment issue. @malfet any idea how to avoid it (https://github.com/pytorch/pytorch/actions/runs/8092102563/job/22113262294?pr=117475)?

Shall I skip this one for python 3.12+ ?

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
+  auto N = B.size(0);
+  auto K = A.size(1);
+
+  TORCH_CHECK(A.dtype() == kBFloat16,
+      "_weight_int4pack_mm: expect A to be bfloat16 tensor.");
```
Why not float16 or float32?

mingfeima replied:
gpt-fast right now is strictly limited to A (bfloat16) * B (int4 or int8) = C(bfloat16), so float16 and float32 are not used. The CUDA counterpart also supports only bfloat16.

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(B.size(1) == K / 2);
+
+  TORCH_CHECK(
+      qGroupSize == 32 || qGroupSize == 64 || qGroupSize == 128 ||
+      qGroupSize == 256);
```
Nit (also, consider adding some user readable error message (and what is wrong with say 512 groupsize)
```suggestion
      qGroupSize >= 32 && qGroupSize <= 256 || (qGroupSize & (qGroupSize - 1)) == 0);
```

mingfeima replied:
the frontend of gpt-fast has constrait for qGroupSize to be 32, 64, 128, 256 now : https://github.com/pytorch-labs/gpt-fast/blob/main/quantize.py#L402. 

larger group size has smaller model size (but the benefit is not big since `ScalesAndZeros` is already much smaller compared to the int4 packed weight) but will have worse accuracy. (group size has little impact on perf)

by default gpt-fast uses group size of 32 which is the same as ggml.

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
+      qGroupSize == 32 || qGroupSize == 64 || qGroupSize == 128 ||
+      qGroupSize == 256);
+
+  TORCH_CHECK(qScaleAndZeros.dim() == 3);
+  TORCH_CHECK(qScaleAndZeros.size(1) == N);
+  TORCH_CHECK(qScaleAndZeros.size(2) == 2);
```
Combine in one check and say something like "unexpected qScaleAndZeros dimensions"

mingfeima replied:
updated

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
+  TORCH_CHECK(A.dtype() == kBFloat16,
+      "_weight_int4pack_mm: expect A to be bfloat16 tensor.");
+  TORCH_CHECK(A.is_contiguous(),
+      "_weight_int4pack_mm: expect A to be contiguous.");
```
Nit (use `__func__` rather than repeating function name)
```suggestion
      __func__, ": expect A to be contiguous.");
```

mingfeima replied:
updated

malfet commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3384,5 +3387,69 @@ Tensor kron(const Tensor& self, const Tensor& other) {
+      "_weight_int4pack_mm: expect B to be uint8 tensor.");
+  TORCH_CHECK(B.is_contiguous(),
+      "_weight_int4pack_mm: expect B to be contiguous.");
+  TORCH_CHECK(B.size(1) == K / 2);
```
Add human readable error here

mingfeima replied:
updated

malfet commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,599 @@
+      float c_val = 0;
+      for (const auto k : c10::irange(K)) {
+        int kb = k / BLOCK_K;
+        float scale = static_cast<float>(ScaleAndZeros[kb * ldc * 2 + n * 2]);
+        float zero = static_cast<float>(ScaleAndZeros[kb * ldc * 2 + n * 2 + 1]);
+        float a_val = static_cast<float>(A[m * lda + k]);
```
Nit
```suggestion
        const auto scale = static_cast<float>(ScaleAndZeros[kb * ldc * 2 + n * 2]);
        const auto zero = static_cast<float>(ScaleAndZeros[kb * ldc * 2 + n * 2 + 1]);
        const auto a_val = static_cast<float>(A[m * lda + k]);
```

mingfeima replied:
updated!

malfet commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,599 @@
+#elif defined(CPU_CAPABILITY_AVX2) && !defined(_MSC_VER)
+        if (nb_size == BLOCK_N) {
+          // for nb_size 32
+          for (int d = 0; d < 16; d++) {
```
```suggestion
          for (const auto d: c10::irange(16)) {
```

mingfeima replied:
updated!

malfet commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,599 @@
+    int64_t qGroupSize,
+    const Tensor& qScaleAndZeros) {
+
+  const BFloat16* A_data = A.data_ptr<BFloat16>();
+  const uint8_t* B_data = B.data_ptr<uint8_t>();
+  BFloat16* C_data = C.data_ptr<BFloat16>();
+  const BFloat16* S_data = qScaleAndZeros.data_ptr<BFloat16>();
```
```suggestion
  const auto* A_data = A.data_ptr<BFloat16>();
  const auto* B_data = B.data_ptr<uint8_t>();
  auto* C_data = C.data_ptr<BFloat16>();
  const auto* S_data = qScaleAndZeros.data_ptr<BFloat16>();
```

mingfeima replied:
updated

malfet commented on aten/src/ATen/native/cpu/utils.h:
```diff
@@ -53,6 +53,31 @@ inline bool data_index_step(T& x, const T& X, Args&&... args) {
   return false;
 }
 
+// Helper struct for compile time loop unrolling
+template <int i>
+struct compile_time_for {
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {
+    compile_time_for<i - 1>::op(function, args...);
+    function(std::integral_constant<int, i - 1>{}, args...);
+  }
+};
+
+template <>
+struct compile_time_for<1> {
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {
+    function(std::integral_constant<int, 0>{}, args...);
+  }
+};
+
+template <>
+struct compile_time_for<0> {
+  // 0 loops, do nothing
+  template <typename Lambda, typename... Args>
+  inline static void op(const Lambda& function, Args... args) {}
+};
+
```
Isn't it identical to https://github.com/pytorch/pytorch/blob/9b2c35b4fe933992a1d7c9ea1765b25f2f64e35f/c10/util/Unroll.h#L13-L19

mingfeima replied:
Ohhhh! did not realize `ForcedUnroll`, many thanks !

malfet replied:
Not sure if it's doing the same, but if it does can be reused (or updated to C++17 goodness)

mingfeima replied:
it does the same thing! But i need to update `c10::ForcedUnroll` a little bit here.  So as to add `Args...`. Also have to make `i` as compile time constant, so the compiler can do `constexpr  int row = i / COLS;` and `constexpr  int col = i % COLS;`  for us. Integer div is very slow on CPU.

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "fails meta-internal tests" -c ghfirst

@mingfeima, looks like this PR breaks meta-internal int4 quant tests. I can't share the whole context, but here's some excerpts:
https://gist.github.com/izaitsevfb/0d55264b03d1472968f3114e1db3f100

```
======================================================================
ERROR: test_save_load_int4woqtensors (pytorch.ao.test.test.TestSaveLoadMeta)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)

...

  File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/torch/__init__.py", line 1121, in _check_with
    raise error_type(message_evaluated)
torch._dynamo.exc.TorchRuntimeError: Failed running call_module L__self___lin1(*(FakeTensor(..., device='cuda:0', size=(32, 64), dtype=torch.bfloat16),), **{}):
shape '[32, 32]' is invalid for input of size 128

from user code:
   File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/pytorch/ao/test/test.py", line 1026, in forward
    x = self.lin1(x)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

cc @malfet
internal: [D54443571](https://www.internalfb.com/diff/D54443571)

mingfeima commented:
> @pytorchbot revert -m "fails meta-internal tests" -c ghfirst
> 
> @mingfeima, looks like this PR breaks meta-internal int4 quant tests. I can't share the whole context, but here's some excerpts: https://gist.github.com/izaitsevfb/0d55264b03d1472968f3114e1db3f100
> 
> ```
> ======================================================================
> ERROR: test_save_load_int4woqtensors (pytorch.ao.test.test.TestSaveLoadMeta)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/torch/utils/_contextlib.py", line 115, in decorate_context
>     return func(*args, **kwargs)
> 
> ...
> 
>   File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/torch/__init__.py", line 1121, in _check_with
>     raise error_type(message_evaluated)
> torch._dynamo.exc.TorchRuntimeError: Failed running call_module L__self___lin1(*(FakeTensor(..., device='cuda:0', size=(32, 64), dtype=torch.bfloat16),), **{}):
> shape '[32, 32]' is invalid for input of size 128
> 
> from user code:
>    File "/re_cwd/buck-out/v2/gen/fbcode/299b386e280e5808/pytorch/ao/test/__torchao_tests__/torchao_tests#link-tree/pytorch/ao/test/test.py", line 1026, in forward
>     x = self.lin1(x)
> 
> Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
> ```

@izaitsevfb it feels like dynamo compiles with the wrong shape. Given a weight of {32, 64} where n=32 and k=64, it will be {n, k/2} with CPU device and { n // 8, k // (inner_k_tiles * 16), 32, inner_k_tiles // 2} on CUDA. I think we can start from `change_linear_weights_to_int4_woqtensors` to debug this issue.

dervon commented:
@mingfeima, It seems that the behavier of cpu and gpu are different for op _convert_weight_to_int4pack.

# on cpu
```python
>>>import torch
>>>torch.__version__
'2.3.0a0+gitf84375c'
>>>torch._convert_weight_to_int4pack(torch.arange(16*16, dtype=torch.int32).reshape(16, 16), 8).shape
torch.Size([16, 8])
```

# on gpu
```python
>>>import torch
>>>torch.__version__
2.2.1+cu121
>>>torch._convert_weight_to_int4pack(torch.arange(128*256, dtype=torch.int32).reshape(128, 256).cuda(), 8).shape
torch.Size([16, 2, 32, 4])
```

As you can see, 
the dim of output are different.

mingfeima commented:
> @mingfeima, It seems that the behavier of cpu and gpu are different for op _convert_weight_to_int4pack.
> 

Yes, cpu and gpu has different packed formats for int4. cpu uses a 2d tensor and gpu uses a 4d tensor. Will this be a trouble for dynamo? Will the issue be fixed if I use a fake 4d tensor, e.g. {n, k/2, 1, 1} for cpu packed weight ?

malfet commented:
@mingfeima sorry, I've missed this part earlier, but point of meta registration is that for the same input shape it should return some output shapes, regardless of whether tensor is on CPU, GPU, XPU or whatever. How interpret this output is up to operator implementation, but shapes and dtypes be the same. So, if you don't mind, can you change the code to return 4D int32(each int32 is a pack of 4 int8) tensor rather than 2D int8 tensor?

mingfeima committed:
Update on "add int4 packed gemm support on CPU device"


This patch adds int4 packed gemm support on CPU, both `avx512` and `avx2` are supported. It is used to speedup https://github.com/pytorch-labs/gpt-fast

The default perf measured on Intel (R) Xeon (R) CPU Max 9480, single socket (56 cores) is `16.13 sec total, 12.40 tokens/sec`

* WOQ int4 on avx512: `5.92 sec total, 33.79 tokens/sec`
* WOQ int4 on avx2: `6.90 sec total, 29.00 tokens/sec`

WOQ int4 is measured with method: https://github.com/pytorch-labs/gpt-fast?tab=readme-ov-file#int4-weight-only-quantization

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@malfet Just updated this patch to use the same shape for packed weight on both CPU and CUDA device! Hopefully dynamo tests won't fail this time.

malfet commented:
> @mingfeima, looks like this PR breaks meta-internal int4 quant tests. I can't share the whole context, but here's some excerpts: https://gist.github.com/izaitsevfb/0d55264b03d1472968f3114e1db3f100

@izaitsevfb just FYI, [torch.ao](https://github.com/pytorch-labs/ao) is not internal only, its just that we run those tests right away internally, but would've taken us a day to discover it in OSS, when torch.ao will pick latest nightly

Test that was failing is likely this one https://github.com/pytorch-labs/ao/blob/c9b397de3895610cfbbca2ccef96fc12c9208885/test/test.py#L1013

malfet commented:
@pytorchbot merge

malfet commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -0,0 +1,597 @@
+        }
+#else
+        for (int n = 0; n < nb_size; n += 2) {
+          int32_t val0 = src[n * K + k];
```
I'm curious, what is the reason for transposing the tensor while packing it?

mingfeima replied:
The major reason is that x86 doesn't have a horizontal reduce from a SIMD vector to a Scalar. Note that on avx512, `_mm512_reduce_add_ps` is a SEQUENCE instruction, which means it will be translated into multiple uops (6 or 7 I don't remember precisely).

If we do gemm in the form of NT (which is default in pytorch Linear definition): {M, K} * {N, K}, we will end up doing dot-product on each m row by n col, and eventually we have a horizontal reduce and write to C would be scalar.

A better approach would be NN： {M, K} * {K, N},  we are doing FMA, broadcast each A[row_index] to a vector and load 4 or 6 vectors from B cols, then use C as an accumulator, we won't need a horizontal reduce and write to C would be a vectorized store.

Above is the explanation of avx512f with fp32 FMA, the other ISAs are similar:
* For avx512f, we transpose B from [N, K] to [K, N]
* For avx-vnni, we can assume that 4 * int8 is a float32, so we packed B from [N, K] to [K/4, N, 4], e.g. [K/4, N4]
* For avx512-bf16 or avx512-f16, we can assume that 2 * bf16 or 2 * f16 is a float32, so we packed B from [N, K] to [K/2, N, 2], e.g. [K/2, N2]
* For amx-vnni, it is just doing 16 cols together, you may assume that [K/4, N4] = [K/4, 64]
* for amx-bf16 or amx-f16, also it it doing 16 cols together, you may assume that [K/2, N2] = [k/2, 32]

So you see, all the gemms (fp32, int8, bf16) are following the same pattern.

The above is the explanation of the inner block, usually we need to break the weight B into multiples of inner blocks, this is ISA dependent:

For this piece of code, on avx512 we have 32 regs and we want to use them efficiently. A will take 1 reg, B will take 4, and C will take 4 by 4, the zero and scale takes 2 by 4. So we end up having block size of 4 x 64 (4 x 4 x 16 floats). avx512-vnni, avx512-bf16 follow the same rule. avx2 would be different as it has only 16 regs.

For amx, we have 8 tiles. Usually, the pattern is 2-2-4 or 1-3-3, which means A takes 2 tiles and B takes 2 tiles, C takes 4 tiles. and each tile computes 16 rows and 16 cols, so the inner block size is usually 16x16 and we handle 2 A blocks and 2 B blocks at a tiles (the 2-2-4 pattern). When M < 16, we use 1-3-3 pattern, waste 1 tile here.

It is often very difficult to understand the packed layout if we read from oneDNN primitive verbose. We can break it down into 2 parts and in this way it would be much easier to sort out the layout:
* the inner block (described in the front), are all the same no matter int8, bf16, f16, f32;
* the outer block (described in the latter), are related to register allocation and caching behaviors.
